File: firmware/src/dsp/normalizer.c
#include "dsp/normalizer.h"

#include <math.h>

void lewis_zscore_normalize(float* buffer, size_t len)
{
    if (!buffer || len == 0) {
        return;
    }

    /* Primeira passada: media. */
    float sum = 0.0f;
    for (size_t i = 0; i < len; ++i) {
        sum += buffer[i];
    }
    float mean = sum / (float)len;

    /* Segunda passada: desvio padrao (populacional). */
    float sq_sum = 0.0f;
    for (size_t i = 0; i < len; ++i) {
        float diff = buffer[i] - mean;
        sq_sum += diff * diff;
    }
    float std = sqrtf(sq_sum / (float)len);

    /* Normaliza in-place. */
    if (std > 0.0f) {
        for (size_t i = 0; i < len; ++i) {
            buffer[i] = (buffer[i] - mean) / std;
        }
    } else {
        for (size_t i = 0; i < len; ++i) {
            buffer[i] = 0.0f;
        }
    }
}

File: tests/fixtures/normalizer.py
"""Normalizacao de sinal para o pipeline DSP do firmware."""

from __future__ import annotations

import numpy as np


def zscore_normalize(x: np.ndarray) -> np.ndarray:
    """Aplica Z-score em um array float32 (equivalente a lewis_zscore_normalize).

    Reproduz a implementacao C em precisao simples: media em uma passada,
    desvio padrao populacional em segunda passada e normalizacao in-place.
    """
    x = x.astype(np.float32)
    n = float(x.shape[0])
    mean = np.sum(x) / np.float32(n)
    diff = x - mean
    std = np.sqrt(np.sum(diff * diff) / np.float32(n))
    if std == 0.0:
        return np.zeros_like(x, dtype=np.float32)
    return (diff / std).astype(np.float32)
